rust/src/model.rs

```diff
@@ -0,0 +1,569 @@
+use std::convert::TryFrom;
+
+use num::{bigint::BigInt, clamp, rational::Ratio, traits::identities::Zero};
+use rand::SeedableRng;
+use rand_chacha::ChaCha20Rng;
+
+use crate::{
+    mask::{config::MaskConfig, seed::MaskSeed, Integers, MaskedModel},
+    utils::generate_integer,
+    PetError,
+};
+
+/// Masking of models.
+pub trait MaskModels<N> {
+    /// Get a reference to the weights.
+    fn weights(&self) -> &Vec<N>;
+
+    /// Cast the weights as ratios. Must handle non-finite weights.
+    fn as_ratios(&self) -> Vec<Ratio<BigInt>>;
+
+    /// Mask the model wrt the mask configuration. Enforces bounds on the scalar and weights.
+    fn mask(&self, scalar: f64, config: &MaskConfig) -> (MaskSeed, MaskedModel) {
```

Comment:
Could we document how the mask is computed?

Reply:
definitely. i would like to link the document Michael once drafted for that, but this is gone from confluence, not sure why. for now i will write some lines in the docstring. just didn't have time yet.

Reply:
Michael provided the document and i added it to confluence: https://xainag.atlassian.net/wiki/spaces/FP/pages/542408769/Masking